### PR TITLE
Spec reportEvent to custom destination URLs with macro substitution

### PR DESCRIPTION
**Pull request by gtanzer**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/pull/113.html" title="Last updated on Sep 26, 2023, 4:20 PM UTC (29bc781)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/113/3b93e20...29bc781.html" title="Last updated on Sep 26, 2023, 4:20 PM UTC (29bc781)">Diff</a>

### CODE REVIEW

File: spec.bs
Comment (gtanzer) [inline spec.bs:804]: @domfarolino Do you know of any existing spec for functionality like this? (take a map of string -> string, and substitute the keys -> values in another string without recursive substitution)

File: spec.bs
Comment (qingxinwu) [inline spec.bs:795]: If ...

File: spec.bs
Comment (gtanzer) [reply spec.bs:870]: Oops, thanks for catching

File: spec.bs
Comment (gtanzer) [reply spec.bs:742]: might as well

File: spec.bs
Comment (gtanzer) [reply spec.bs:1323]: surprised it didn't give me a compile error

File: spec.bs
Comment (gtanzer) [reply spec.bs:795]: Fixed

Comment (gtanzer) [reply spec.bs:795]: Fixed.

**Comment by gtanzer** [comment]
@qingxinwu Thanks for reviewing! Fixed all the comments.

File: spec.bs
Comment (domfarolino) [reply spec.bs:805]: Hmm, the only kind of substitution like this that comes to mind is when we substitute format specifiers in e.g., `console.log('My number is: %f', .435)`, which is in https://console.spec.whatwg.org/#formatter. That's a pretty manual process, and I don't know of any preexisting reusable utility for this unfortunately, so you'll probably just have to do it that way.

File: spec.bs
Comment (domfarolino) [inline spec.bs:742]: does this need an initial value, like the boolean below?

File: spec.bs
Comment (domfarolino) [inline spec.bs:1272]: Can you add tighter line breaks here to accommodate the narrow IDL viewing window in the spec preview? (Same with below).

File: spec.bs
Comment (domfarolino) [inline spec.bs:1270]: I can't remember, is there a technical reason why we can't express this as a [union type](https://webidl.spec.whatwg.org/#dfn-union-type), and instead have to express this as a dictionary with these two implicit modes? If there is we should document it here, and if there is not we should try and follow-up with a web-compatible change to make this more idiomatic and less error-prone.

Comment (gtanzer) [reply spec.bs:1270]: It was due to distinguishability. The issue isn't as clear in the spec because reportEvent for component ads hasn't been spec'd yet, which makes some of the required fields optional.

File: spec.bs
Comment (gtanzer) [reply spec.bs:742]: I think it's always specified, but might as well give it a default.

File: spec.bs
Comment (gtanzer) [reply spec.bs:805]: Factored it out into a TODO. The exact semantics of this are not super important right now.

**Comment by gtanzer** [comment]
To be merged after https://github.com/WICG/fenced-frame/pull/117 , which creates [=fencedframeutil/substitute macros=].

**Comment by domfarolino** [comment]
> To be merged after https://github.com/WICG/fenced-frame/pull/117 , which creates [=fencedframeutil/substitute macros=].

Sounds good, I think this mostly looks good so let's wait until that gets landed then.

**Review by domfarolino — APPROVED** [review]
@qingxinwu Does this look good to you?

File: spec.bs
Comment (qingxinwu) [inline spec.bs:751]: nit: it would be good to briefly explain what this field means, and what null means for the field.

File: spec.bs
Comment (qingxinwu) [inline spec.bs:894]: IIRC, the implementation also checks if the event's origin is enrolled. Are we assuming that this same-origin check with `allowed reporting origins` covers that? If so, would be good to document at [line 750](https://github.com/WICG/fenced-frame/pull/113/files#diff-6f5a1d8263b0b0c42e2716ba5750e3652e359532647ac934c1c70086ae3ceddaR750) that all origins in [=fenced frame reporting metadata/allowed reporting origins=] are enrolled.

File: spec.bs
Comment (qingxinwu) [inline spec.bs:1455]: nit: For each.  s/f/F

File: spec.bs
Comment (qingxinwu) [inline spec.bs:747]: I don't find where this field is set. It seems it always uses the default `true`. Is that expected?

File: spec.bs
Comment (qingxinwu) [reply spec.bs:751]: or maybe just say "passed from protected audience interest group's allowed reporting origins", with link to https://github.com/WICG/turtledove/blob/main/spec.bs#L3201.

File: spec.bs
Comment (gtanzer) [reply spec.bs:747]: It's set here:

https://wicg.github.io/turtledove/#:~:text=Set%20pendingConfig%E2%80%99s%20fenced%20frame%20reporting%20metadata%20to%20a%20struct%20with%20the%20following%20items%3A

File: spec.bs
Comment (gtanzer) [reply spec.bs:894]: Attestation checks are performed when sending the beacon, i.e. it uses an already present check on line 823.

File: spec.bs
Comment (qingxinwu) [reply spec.bs:747]: oh I see. Missed that, because clicking on the field cannot show usages like this. Should have done a search.
feel free to close this comment then.

File: spec.bs
Comment (qingxinwu) [reply spec.bs:894]: Ack. Forget about the "interesting" `optionally` part.

**Comment by qingxinwu** [comment]
LGTM.

File: spec.bs
Comment (gtanzer) [reply spec.bs]: Added "An origin must be present in this list to be the destination of a [=fencedframetype/destination URL event=] report."

**Pull request merged**
